src/server/debug/server/server.go

```diff
@@ -713,43 +714,59 @@ func (s *debugServer) getWorkerPodsLoki(pipelineInfo *pps.PipelineInfo) (map[str
 	return pods, nil
 }
 
-// TODO: Need to bump up this default on the server side, it is too low as is.
-const maxLogs = 5000
+// This used to be 5,000 but a comment said this was too few logs, so now it's 30,000.  If it still
+// seems too small, bump it up again.  5,000 remains the maximum value of "limit" in queries that
+// Loki seems to accept.
+const (
+	maxLogs       = 30000
+	serverMaxLogs = 5000
+)
 
-func (s *debugServer) queryLoki(queryStr string, cb func(loki.LabelSet, string) error) error {
+func (s *debugServer) queryLoki(ctx context.Context, queryStr string, cb func(loki.LabelSet, string) error) error {
 	c, err := s.env.GetLokiClient()
 	if err != nil {
-		return errors.EnsureStack(err)
+		return errors.EnsureStack(errors.Errorf("get loki client: %v", err))
 	}
-	start := time.Now().Add(-(30 * 24 * time.Hour))
-	end := time.Now()
-	for {
-		// TODO: Need a real context.
-		resp, err := c.QueryRange(s.env.Context(), queryStr, maxLogs, start, end, "FORWARD", 0, 0, true)
+	for numLogs, start, end := 0, time.Now().Add(-(30 * 24 * time.Hour)), time.Now(); start.Before(end) && numLogs < maxLogs; {
+		resp, err := c.QueryRange(ctx, queryStr, serverMaxLogs, start, end, "FORWARD", 0, 0, true)
 		if err != nil {
-			return err
+			// The error from QueryRange has a stack.
+			return errors.Errorf("query range (query=%v, maxLogs=%v, start=%v, end=%v): %+v", queryStr, maxLogs, start.Format(time.RFC3339), end.Format(time.RFC3339), err)
 		}
 		streams, ok := resp.Data.Result.(loki.Streams)
 		if !ok {
 			return errors.Errorf("resp.Data.Result must be of type loki.Streams")
 		}
-		var numLogs int
+		var advancedStart bool
+		var readThisIteration int
 		for _, stream := range streams {
 			for _, entry := range stream.Entries {
 				numLogs++
+				readThisIteration++
 				if entry.Timestamp.After(start) {
+					advancedStart = true
 					start = entry.Timestamp
 				}
-				if err := cb(stream.Labels, entry.Line); err != nil {
+				if err := cb(stream.Labels, entry.Line+"\n"); err != nil {
```

Comment:
Rather than `queryLoki` adding a newline to every line, I think that the responsibility should be on the callback to handle the line appropriately, e.g. quoting it in JSON or writing terminated with a newline in a log file, or whatever.

Looking at where `queryLoki` is called, it looks like `collectLogsLoki` is the one place that actually uses the line. Rather than `w.Write([]byte(line))`, it could just call `fmt.Fprintln(w, line)`.

Reply:
Done.